plaidml/bridge/openvino/ops/deformable_convolution.cpp

```diff
@@ -0,0 +1,284 @@
+    std::pair<size_t, size_t> result(pad_before, output_size);
+    return result;
+  }
+  THROW_IE_EXCEPTION << "Unexpected autopadding mode.";
+}
+
+// Extract values which are needed.
+edsl::Tensor extract_tensor(edsl::Tensor I, int rank) {
+  std::vector<TensorDim> I_dims((rank - 2) * (rank + 1) + 2), O_dims(rank);
+  std::vector<TensorIndex> I_idxs((rank - 2) * (rank + 1) + 2), O_idxs(rank);
+  I.bind_dims(I_dims);
+  // Set up the dimensions of output.
+  for (auto i = 2; i < rank + 2; ++i) {
+    O_dims[i - 2] = I_dims[i];
+  }
+  // Set up the indexs between input and output.
+  for (auto i = 2; i < rank + 2; ++i) {
+    I_idxs[i] = O_idxs[i - 2];
+  }
+  // Set up the batch_size indexs of input.
+  for (auto i = 0; i < 2; ++i) {
+    I_idxs[i] = I_idxs[i + 2];
+  }
+  for (auto i = 1; i < rank - 2; ++i) {
+    for (auto j = 0; j < rank; ++j) {
+      I_idxs[i * (rank + 1) + 2 + j] = I_idxs[(i - 1) * (rank + 1) + 2 + j];
+    }
+  }
+  // Set up the hight and width of index.
+  for (auto i = 0; i < rank - 2; ++i) {
+    I_idxs[(i + 1) * (rank + 1) + 1] = edsl::TensorIndex(i);
+  }
+  edsl::Tensor O = Contraction(O_dims, O_idxs).assign(I(I_idxs));
+  return O;
+}
+
+// Cast long int to int in vector.
+std::vector<int> cast_vector(std::vector<int64_t> vec) {
+  std::vector<int> cast_vec(vec.size());
+  for (auto i = 0; i < vec.size(); ++i) {
+    cast_vec[i] = static_cast<int>(vec[i]);
+  }
+  return cast_vec;
+}
+
+// Compute DeformableConvolution.
+edsl::Tensor compute_deformable_convolution(edsl::Tensor I, edsl::Tensor OFF, edsl::Tensor F, std::vector<int> I_shape,
+                                            std::vector<int> OFF_shape, std::vector<int> F_shape, int G, int DG,
+                                            size_t rank, std::vector<size_t> strides, std::vector<size_t> dilations,
+                                            std::vector<size_t> pad_befores, plaidml::op::AutoPadMode autopad_mode) {
+  // Define dim of each tensor.
+  int N = I_shape[0];
+  int CI = I_shape[1];
+  int CO = F_shape[1];
+  int OFF_C = OFF_shape[1];
+  // Throw exception
+  if (CI % G != 0 || CO % G != 0 || CI % DG != 0 || OFF_C % DG != 0) {
+    THROW_IE_EXCEPTION << "Incorrected shape for DeformableConvolution.";
+  }
+  // Define new dimention.
+  std::vector<int> NEW_DIM;
+  for (auto i = 2; i < rank; ++i) {
+    NEW_DIM.push_back(F_shape[i] * OFF_shape[i]);
+  }
+  // Set up the value of the reshape to split the dimension.
+  std::vector<int> OFF_reshape_split;
+  OFF_reshape_split.push_back(N);
+  OFF_reshape_split.push_back(DG);
+  OFF_reshape_split.insert(OFF_reshape_split.end(), F_shape.begin() + 2, F_shape.end());
+  OFF_reshape_split.push_back(rank - 2);
+  OFF_reshape_split.insert(OFF_reshape_split.end(), OFF_shape.begin() + 2, OFF_shape.end());
+  edsl::Tensor offset = op::reshape(OFF, make_tuple<int>(OFF_reshape_split));
+  // Set up the value of transpose.
+  std::vector<int> OFF_transpose;
```

Comment:
Please use `OFF_transpose_dims`

Reply:
I think adding some comments here for what each dimention is would be better for readers to understand.